File: tests/unit/test_shell_icon_customize.mjs
/*
 * Behavioural test for desktop-icon CUSTOMIZATION (glyph / label / color).
 *
 * Drives the REAL integrations/agent_engine/static/hartDesktop.js through its
 * public API on a tiny dependency-free DOM shim (CI runners here have no jsdom):
 *   1. load the real module  -> init() renders the default icons
 *   2. window.hartCustomizeIcon('feed')  -> builds the real dialog, seeds inputs
 *   3. type a Material glyph + label + color, click Save
 *   4. assert the icon ELEMENT updated (observable DOM)   AND
 *      HartSession.set('desktop_icons', …) persisted the override (side-effect)
 *   5. repeat with an EMOJI glyph -> asserts it is NOT wrapped in the icon font
 *   6. Reset -> override cleared from both the DOM and the persisted blob
 *
 * Run:  node tests/unit/test_shell_icon_customize.mjs
 * (A Python wrapper, test_shell_icon_customize.py, shells out to this so the
 *  suite/CI picks it up via pytest too.)
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import vm from 'node:vm';

const HERE = dirname(fileURLToPath(import.meta.url));
const STATIC = join(HERE, '..', '..', 'integrations', 'agent_engine', 'static');
const SRC_BRAND = join(STATIC, 'hartBrandArt.js');   // shared brand-art (glyph + gradient)
const SRC = join(STATIC, 'hartDesktop.js');

let failures = 0;
function ok(cond, msg) { if (cond) { console.log('  OK   ' + msg); } else { failures++; console.log(' FAIL  ' + msg); } }
function eq(a, b, msg) { ok(a === b, msg + '  (got ' + JSON.stringify(a) + ', want ' + JSON.stringify(b) + ')'); }

// ── Minimal DOM shim ───────────────────────────────────────────────────────
// Real enough to run the module: elements track attributes/style/children, and
// innerHTML resets children (the code re-queries known selectors afterwards, so
// querySelector lazily materializes a child stub per selector — no HTML parser).
function makeEl(tag) {
  const el = {
    tagName: (tag || 'div').toUpperCase(),
    _attrs: {}, _kids: [], _bySel: {}, _listeners: {},
    style: {}, classList: { _s: new Set(),
      add(c) { this._s.add(c); }, remove(c) { this._s.delete(c); },
      contains(c) { return this._s.has(c); } },
    textContent: '', _innerHTML: '', parentNode: null, pointerType: '',
    get innerHTML() { return this._innerHTML; },
    set innerHTML(v) { this._innerHTML = v; this._kids = []; this._bySel = {}; },
    setAttribute(k, v) { this._attrs[k] = String(v); },
    getAttribute(k) { return Object.prototype.hasOwnProperty.call(this._attrs, k) ? this._attrs[k] : null; },
    removeAttribute(k) { delete this._attrs[k]; },
    hasAttribute(k) { return Object.prototype.hasOwnProperty.call(this._attrs, k); },
    appendChild(c) { c.parentNode = el; this._kids.push(c); return c; },
    removeChild(c) { this._kids = this._kids.filter(k => k !== c); c.parentNode = null; return c; },
    addEventListener(t, fn) { (this._listeners[t] = this._listeners[t] || []).push(fn); },
    removeEventListener() {},
    dispatch(t, ev) { (this._listeners[t] || []).forEach(fn => fn(ev || { target: el, preventDefault() {}, stopPropagation() {} })); },
    focus() {}, select() {}, setPointerCapture() {}, releasePointerCapture() {},
    closest() { return null; },
    querySelector(sel) {
      // exact-id match against real children first
      for (const k of this._kids) if (k._attrs.id && ('#' + k._attrs.id) === sel) return k;
      // `.desktop-icon[data-id="X"]` — match a real child by its data-id
      const m = /^\.desktop-icon\[data-id="(.+)"\]$/.exec(sel);
      if (m) return this._kids.find(k => k._attrs['data-id'] === m[1] &&
                                          (k._attrs.class || '').split(' ').includes('desktop-icon')) || null;
      if (!this._bySel[sel]) { const s = makeEl('div'); s.parentNode = el; this._bySel[sel] = s; }
      return this._bySel[sel];
    },
    querySelectorAll(sel) {
      if (sel === '.desktop-icon') return this._kids.filter(k => (k._attrs.class || '').split(' ').includes('desktop-icon'));
      if (sel === '.desktop-icon.selected') return this._kids.filter(k => k.classList.contains('selected'));
      return [];
    },
    get className() { return this._attrs.class || ''; },
    set className(v) { this._attrs.class = v; },
    // Real DOM reflects the `id` / `dataset`-free property to the attribute.
    get id() { return this._attrs.id || ''; },
    set id(v) { this._attrs.id = String(v); }
  };
  return el;
}

const registry = {};   // id -> element (so getElementById finds layer + ctx-menu)
const document = {
  readyState: 'complete',
  documentElement: makeEl('html'),
  body: makeEl('body'),
  _listeners: {},
  createElement: (t) => makeEl(t),
  getElementById(id) {
    if (registry[id]) return registry[id];
    // also search the live tree (dialogs are appended to <body>, not registered)
    const walk = (n) => { for (const k of n._kids) { if (k._attrs.id === id) return k; const r = walk(k); if (r) return r; } return null; };
    return walk(this.body);
  },
  addEventListener(t, fn) { (this._listeners[t] = this._listeners[t] || []).push(fn); }
};
const layer = makeEl('div'); layer.setAttribute('id', 'hart-desktop'); registry['hart-desktop'] = layer;
const ctxMenu = makeEl('div'); ctxMenu.setAttribute('id', 'ctx-menu'); registry['ctx-menu'] = ctxMenu;
document.body.appendChild(layer);

// Real-ish HartSession: capture every persisted blob (the side-effect we assert).
let lastSaved = null;
const sessionBlob = {};
const sandbox = {
  document, console,
  setTimeout: (fn) => { fn(); return 0; },   // run deferred work inline
  clearTimeout() {}, requestAnimationFrame() { return 0; }, cancelAnimationFrame() {},
  getComputedStyle: () => ({ getPropertyValue: () => '#6c63ff' }),
  // `appearance` carries a manifest-stamped colour (shell_manifest.with_icon_colors)
  // to exercise the de-monochrome path: it should RENDER coloured but a plain Save
  // must NOT persist it as a per-icon override (blob stays lean).
  MANIFEST: { feed: { title: 'Feed', icon: 'rss_feed' }, recipes: { title: 'Recipes', icon: 'menu_book' },
              appearance: { title: 'Appearance', icon: 'palette', color: '#22cc88' } },
  ctxItem: (icon, label, action) => '<i>' + icon + '|' + label + '|' + action + '</i>',
  ctxSep: () => '<sep>',
  openPanel() {},
  HartSession: {
    ready(cb) { cb(sessionBlob); },
    get(k, d) { return Object.prototype.hasOwnProperty.call(sessionBlob, k) ? sessionBlob[k] : d; },
    set(k, v) { sessionBlob[k] = v; if (k === 'desktop_icons') lastSaved = v; }
  }
};
sandbox.window = sandbox;            // classic-script `window` is the global
vm.createContext(sandbox);
// hartDesktop renders glyphs/art tiles through the shared window.HartBrandArt
// (loaded first in the real shell), so define it here before the module runs.
vm.runInContext(readFileSync(SRC_BRAND, 'utf8'), sandbox, { filename: 'hartBrandArt.js' });
vm.runInContext(readFileSync(SRC, 'utf8'), sandbox, { filename: 'hartDesktop.js' });

// ── 0. defaults rendered (feed present) ──
const iconOf = (id) => layer._kids.find(k => k._attrs['data-id'] === id);
ok(iconOf('feed'), 'default icons rendered (feed icon present)');

// ── 1. open the customize dialog for `feed` ──
sandbox.window.hartCustomizeIcon('feed');
const dlg = registry['hart-icon-customize'] || sandbox.document.body._kids.find(k => k._attrs.id === 'hart-icon-customize');
ok(dlg, 'customize dialog appended to <body>');
// the dialog seeds inputs from the MANIFEST default
const gIn = dlg.querySelector('#hic-glyph'), lIn = dlg.querySelector('#hic-label'), cIn = dlg.querySelector('#hic-color');
eq(gIn.value, 'rss_feed', 'glyph input seeded from MANIFEST default');
eq(lIn.value, 'Feed', 'label input seeded from MANIFEST default');

// ── 2. type a Material glyph + label, pick a color, Save ──
gIn.value = 'auto_awesome'; gIn.dispatch('input');
lIn.value = 'My Feed'; lIn.dispatch('input');
cIn.value = '#ff8800'; cIn.dispatch('input');                 // turns the color override on
dlg.querySelector('#hic-save').dispatch('click', { target: dlg.querySelector('#hic-save'), preventDefault() {} });

// observable DOM mutation on the real icon element
const fe = iconOf('feed');
eq(fe.getAttribute('data-ov-glyph'), 'auto_awesome', 'icon stores glyph override on data-attr');
eq(fe.getAttribute('data-ov-label'), 'My Feed', 'icon stores label override on data-attr');
eq(fe.getAttribute('data-ov-color'), '#ff8800', 'icon stores color override on data-attr');
eq(fe.querySelector('.di-label').textContent, 'My Feed', 'label element shows the override');
eq(fe.getAttribute('aria-label'), 'My Feed', 'aria-label updated for screen readers');
// glyphSpan() renders the inner markup as a string into .di-glyph.innerHTML;
// assert against that real output (the shim does not parse HTML into children).
const feGlyphHTML = fe.querySelector('.di-glyph').innerHTML;
ok(/class="mi material-icons-round"/.test(feGlyphHTML), 'Material-name glyph uses the icon font');
ok(feGlyphHTML.indexOf('color:#ff8800') >= 0, 'glyph color applied inline (overrides theme accent)');

// observable persistence side-effect: the override is in the saved blob
ok(Array.isArray(lastSaved), 'persist() wrote desktop_icons');
const saved = lastSaved.find(r => r.id === 'feed');
eq(saved.glyph, 'auto_awesome', 'persisted entry carries glyph override');
eq(saved.label, 'My Feed', 'persisted entry carries label override');
eq(saved.color, '#ff8800', 'persisted entry carries color override');
// un-customized icons stay lean (no override keys)
const recipesSaved = lastSaved.find(r => r.id === 'recipes');
ok(recipesSaved && !('glyph' in recipesSaved) && !('label' in recipesSaved) && !('color' in recipesSaved),
   'un-customized icon serializes to a plain {id,x,y} (no override noise)');

// ── 3. EMOJI glyph is NOT wrapped in the Material icon font ──
sandbox.window.hartCustomizeIcon('recipes');
const dlg2 = sandbox.document.body._kids.find(k => k._attrs.id === 'hart-icon-customize');
const g2 = dlg2.querySelector('#hic-glyph');
g2.value = '🚀'; g2.dispatch('input');
dlg2.querySelector('#hic-save').dispatch('click', { target: dlg2.querySelector('#hic-save'), preventDefault() {} });
const re = iconOf('recipes');
const reGlyphBox = re.querySelector('.di-glyph');
ok(/class="mi di-emoji"/.test(reGlyphBox.innerHTML), 'emoji glyph uses a plain span, not the icon font');
ok(!/material-icons-round/.test(reGlyphBox.innerHTML), 'emoji glyph is NOT wrapped in the icon font');
eq(reGlyphBox.querySelector('.mi').textContent, '🚀', 'emoji rendered as text content');
eq(lastSaved.find(r => r.id === 'recipes').glyph, '🚀', 'emoji override persisted');

// ── 4. Reset clears the override (DOM + persisted) ──
sandbox.window.hartCustomizeIcon('feed');
const dlg3 = sandbox.document.body._kids.find(k => k._attrs.id === 'hart-icon-customize');
dlg3.querySelector('#hic-reset').dispatch('click', { target: dlg3.querySelector('#hic-reset'), preventDefault() {} });
const fe2 = iconOf('feed');
ok(!fe2.hasAttribute('data-ov-glyph') && !fe2.hasAttribute('data-ov-label') && !fe2.hasAttribute('data-ov-color'),
   'Reset clears override data-attrs from the icon');
eq(fe2.querySelector('.di-label').textContent, 'Feed', 'Reset restores the MANIFEST label');
const feSaved = lastSaved.find(r => r.id === 'feed');
ok(feSaved && !('glyph' in feSaved) && !('color' in feSaved), 'Reset removes the override from the persisted blob');

// ── 5. A MANIFEST-stamped colour renders but is NOT persisted as an override ──
// (de-monochrome path: opening Customize on a manifest-coloured icon and Saving
//  without touching the swatch must keep the desktop blob lean.)
const ap = iconOf('appearance');
ok(ap.querySelector('.di-glyph').innerHTML.indexOf('color:#22cc88') >= 0,
   'manifest colour is applied to the glyph on render');
ok(!ap.hasAttribute('data-ov-color'), 'manifest colour is NOT stashed as a per-icon override on render');
sandbox.window.hartCustomizeIcon('appearance');
const dlg4 = sandbox.document.body._kids.find(k => k._attrs.id === 'hart-icon-customize');
dlg4.querySelector('#hic-save').dispatch('click', { target: dlg4.querySelector('#hic-save'), preventDefault() {} });
const apSaved = lastSaved.find(r => r.id === 'appearance');
ok(apSaved && !('color' in apSaved), 'plain Save on a manifest-coloured icon persists NO colour override (lean blob)');

console.log(failures ? ('\nRESULT: ' + failures + ' FAILED') : '\nRESULT: ALL PASS');
process.exit(failures ? 1 : 0);

File: tests/unit/test_shell_icon_customize.py
"""Desktop-icon CUSTOMIZATION (glyph / label / color), macOS-/Windows-style.

Two behavioural layers, no grep-tests:

1. test_icon_customize_behaviour — runs the REAL static/hartDesktop.js through
   its public API on a tiny DOM shim (test_shell_icon_customize.mjs) and asserts
   the full round-trip: open the Customize dialog -> edit glyph/label/color ->
   Save -> the icon ELEMENT mutates AND HartSession persists the override; plus
   emoji-vs-Material glyph rendering and Reset. Skips cleanly if node is absent.

2. test_session_state_roundtrips_icon_overrides — proves the persistence contract
   the feature relies on through the REAL Flask route: a desktop_icons entry that
   carries glyph/label/color survives POST -> GET unchanged (opaque blob), and
   un-customized siblings + unrelated shell state are not clobbered.

Local note: this box OOM-kills the full pytest import chain; the .mjs runs
standalone (`node tests/unit/test_shell_icon_customize.mjs`) and the round-trip
runs through a temp-dir Flask client. Committed for CI (node 20 present there).
"""
import os
import shutil
import subprocess
import sys
import tempfile

import pytest

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
if ROOT not in sys.path:
    sys.path.insert(0, ROOT)

MJS = os.path.join(os.path.dirname(__file__), 'test_shell_icon_customize.mjs')


def _liquid_ui():
    try:
        from integrations.agent_engine.liquid_ui_service import LiquidUIService
    except Exception as e:  # heavy deps absent in a minimal CI runner -> skip
        pytest.skip('LiquidUIService not importable here: ' + str(e))
    return LiquidUIService


def test_icon_customize_behaviour():
    """Drive the REAL hartDesktop.js customize flow end-to-end (Node + DOM shim)."""
    node = shutil.which('node')
    if not node:
        pytest.skip('node not available to run the JS behavioural harness')
    r = subprocess.run([node, MJS], capture_output=True, text=True, timeout=60)
    # Surface the harness' per-assertion log on failure so CI shows exactly which.
    assert r.returncode == 0, 'hartDesktop.js customize harness failed:\n' + r.stdout + r.stderr
    assert 'RESULT: ALL PASS' in r.stdout, r.stdout


def test_session_state_roundtrips_icon_overrides():
    """An icon entry carrying glyph/label/color survives the REAL POST->GET route
    unchanged, and neither un-customized siblings nor other shell state are lost
    (the client read-modify-writes one blob; the backend stores it opaquely)."""
    svc = _liquid_ui()()
    svc._data_dir = tempfile.mkdtemp()  # never touch real shell state
    client = svc._create_flask_app().test_client()

    icons = [
        {'id': 'feed', 'x': 24, 'y': 24,
         'glyph': 'auto_awesome', 'label': 'My Feed', 'color': '#ff8800'},  # customized
        {'id': 'recipes', 'x': 24, 'y': 116},                               # plain
        {'id': 'notes', 'x': 24, 'y': 208, 'glyph': '\U0001f680'},          # emoji glyph only
    ]
    r = client.post('/api/shell/session-state',
                    json={'desktop_icons': icons, 'theme': 'midnight'})
    assert r.status_code == 200

    got = client.get('/api/shell/session-state').get_json()
    assert got.get('desktop_icons') == icons          # overrides + emoji round-trip intact
    assert got.get('theme') == 'midnight'             # unrelated shell state not clobbered


def test_hartdesktop_js_wires_the_customize_action():
    """Source-shape guard (explicitly labelled — paired with the behavioural test
    above, never the only coverage): the context menu offers Customize and the
    handler/applier the menu invokes are defined by name, and customization
    travels in the single desktop_icons writer (no parallel override store)."""
    src = open(os.path.join(ROOT, 'integrations', 'agent_engine', 'static', 'hartDesktop.js'),
               encoding='utf-8').read()
    # WAS: assert "'Customize" in src and "hartCustomizeIcon('" in src
    # Red since at least 2026-07-31 (six consecutive runs of the Shell UI
    # WebKit Safety gate) while the FEATURE WORKED — and wrong in a more
    # instructive way than a plain rename:
    #   * "hartCustomizeIcon('" appears ZERO times. The call sites are
    #     `window.hartCustomizeIcon && window.hartCustomizeIcon(id)` (:345)
    #     and the definition (:389). The literal open-quote could never have
    #     matched this code, so this half was simply wrong.
    #   * "'Customize" DOES still match — but at :401, as
    #     `ov.setAttribute('aria-label', 'Customize icon')` inside the dialog.
    #     The MENU entry it claims to check was renamed to 'Properties'
    #     (:345). So this half was passing for a coincidental reason, in a
    #     different part of the file, about a different thing.
    # A guard half-passing by accident and half-failing on a string that never
    # existed is exactly what feedback_no_grep_tests warns about: it says
    # nothing about whether the wiring survived, in either direction.
    #
    # Assert the WIRING instead: some context-menu entry must invoke the
    # action, and the action must be defined. That is what "menu entry ->
    # action" was always trying to say, and it survives renaming the label.
    assert 'window.hartCustomizeIcon =' in src                    # action defined
    _menu_entry = [ln for ln in src.splitlines()
                   if 'label:' in ln and 'hartCustomizeIcon' in ln]
    assert _menu_entry, (
        "no context-menu entry invokes hartCustomizeIcon — the action is "
        "defined but unreachable from the menu, which is the regression this "
        "guard exists to catch (the LABEL may be renamed freely)")
    assert 'function applyIconVisual' in src                      # single apply path (render + dialog)
    # one writer: overrides ride inside readPositions()/persist(), not a new key
    assert "HartSession.set('desktop_icons'" in src
    assert 'data-ov-glyph' in src                                 # serialized off the DOM, no side store


if __name__ == '__main__':
    # Inline runner (pytest OOMs on this box): execute every test_* and report.
    fns = [v for k, v in sorted(globals().items())
           if k.startswith('test_') and callable(v)]
    failed = 0
    for fn in fns:
        try:
            fn()
            print('  OK  ', fn.__name__)
        except Exception as e:
            failed += 1
            print(' FAIL ', fn.__name__, '->', repr(e))
    print('RESULT:', 'ALL PASS' if not failed else (str(failed) + ' FAILED'))
    sys.exit(1 if failed else 0)
